### PR TITLE
Upstream sessions

### PR DESCRIPTION
**Pull request by jtojnar**
###### Motivation for this change
See https://github.com/NixOS/nixpkgs/issues/39871

This is just a first swallow, in the future pull requests we will want to convert more DEs and investigate Wayland.

###### Things done
- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
- [x] Tested via one or more NixOS tests
     - [x] `gnome3`
     - [x] `gnome3-gdm`
- [x] Tried to run upstream xsession (GNOME) in VM
     - [x] with GDM
     - [x] with lightdm
     - [x] with SDDM
- [x] Tried to run legacy xsession (Xfce) in VM
     - [x] with lightdm
     - [x] with gdm
- [x] Tried to run legacy combined xsession (xmonad+xterm) in VM
     - [x] with lightdm
     - [x] with gdm
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

cc @hedning, @worldofpeace

### CODE REVIEW

**Comment by hedning** [comment]
This seems like good start :)

Did a little testing (with gdm):

#### Login

1. Tested in a VM without password, which worked as expected
2. Tested on hardware with my regular user and couldn't log in

I'm getting this in the log:
```
juli 23 13:10:16 x1 gdm-password][4245]: gkr-pam: no password is available for user
```

Not sure how to reproduce in a VM as the user.password option is ignored for some reason.

#### GPaste

Gnome-control-center can't find the necessary gpaste schema, so it crashes:

1. Started Gnome Shell in a VM (passwordless user) with `services.gnome3.gpaste.enabled = true`. 
2. Turned on the gpaste extension.
3. Tried launching gnome-control-center

Resulting in:
```
Jul 23 11:27:04 nixos .gnome-control-[1908]: Settings schema 'org.gnome.GPaste' is not installed
```

**Comment by jtojnar** [comment]
Add Login, maybe this:

> If the option `users.mutableUsers` is true, the password defined in one of the three options will only be set when the user is created for the first time.

Ad GPaste, that is probably the keyboard shortcuts XML file. I do not think it is possible  Maybe we could create `gnome-control-center-with-plugins` expression. Or just give up and re-add the session path (though I am not sure how to do it).

I tried lightdm now, but it does not even see the session. Also in the log, I see ** (lightdm-gtk-greeter:802): WARNING **: 11:41:49.170: [PIDs] Failed to execute command: /usr/lib/at-spi2-core/at-spi-bus-launcher

**Comment by hedning** [comment]
Right, setting mutable users to false made login work with a hardcoded password in a VM, not sure why it won't work on my hardware though (tested on two machines now, will investigate some more).

**Comment by jtojnar** [comment]
Regarding GPaste, I guess, we also could extend the `KeyListEntries` in the XML file with `schemadir` attribute and patch Control Center to respect it.

File: nixos/modules/services/x11/desktop-managers/gnome3.nix [hunk 1]
```diff
@@ -55,14 +55,6 @@ in {
         description = "Enable Gnome 3 desktop manager.";
       };
 
-      sessionPath = mkOption {
```
Comment (Mic92) [inline nixos/modules/services/x11/desktop-managers/gnome3.nix]: Is there a way to redirect users to the new option (extraSessionFilePackages?) instead?

Comment (jtojnar) [reply nixos/modules/services/x11/desktop-managers/gnome3.nix]: It serves a completely different purpose. The replacement to `sessionPath` is fixing packages.

**Comment by hedning** [comment]
(had a bad `.xprofile` which caused the crash).

Another option for GPaste would be to compile the schema in `/run/current-system/sw/share/glib-2.0/...`, not sure if that's possible in a module though?

**Comment by jtojnar** [comment]
Yep, doing as you suggest in https://github.com/NixOS/nixpkgs/issues/42176#issuecomment-399386269 would be another possibility.

**Comment by worldofpeace** [comment]
Hmm, sddm doesn't see the session either. Maybe that's missing something.

**Comment by worldofpeace** [comment]
This should be fine
```diff
diff --git a/pkgs/desktops/gnome-3/core/gnome-session/default.nix b/pkgs/desktops/gnome-3/core/gnome-session/default.nix
index 7b407aad506..6f9cc384fa2 100644
--- a/pkgs/desktops/gnome-3/core/gnome-session/default.nix
+++ b/pkgs/desktops/gnome-3/core/gnome-session/default.nix
@@ -33,8 +33,10 @@ stdenv.mkDerivation rec {
   preFixup = ''
     for desktopFile in $(grep -rl "Exec=gnome-session" $out/share)
     do
-      echo "Patching gnome-session path in: $desktopFile"
+      echo "Patching Exec gnome-session path in: $desktopFile"
       sed -i "s,^Exec=gnome-session,Exec=$out/bin/gnome-session," $desktopFile
+      echo "Patching TryExec gnome-sesison path in: $desktopFile"
+      sed -i "s,^TryExec=gnome-session,TryExec=$out/bin/gnome-session," $desktopFile
     done
     wrapProgram "$out/bin/gnome-session" \
       --prefix PATH : "${glib.dev}/bin" \

```

**Comment by worldofpeace** [comment]
Also,
```nix
debug = mkEnableOption "gnome-session debug messages";
```

It would be nice to still have an interface to do that.

**Comment by worldofpeace** [comment]
Do these lightdm [errors](https://gist.github.com/worldofpeace/bec0702389e6f23c8670ded86c7233d2) look of any interest?

**Comment by jtojnar** [comment]
Not particularly.

With the latest commit, I am able to launch GNOME Shell in lightdm in VM, but it does not seem to receive environment variables (including `PATH`).

**Review by worldofpeace — COMMENTED** [review]
Ahh so that's what it was about.

**Comment by hedning** [comment]
I took a closer look at how LightDM and NixOS sets up the environment and launches sessions. I'll try to lay out things here for some hopefully shared clarity.

#### The current model

LightDM launches sessions like this ([the same thing goes for sddm](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/sddm.nix#L52)):

1.It extracts the `Exec` command of the selected .desktop session file
2. It passes the command to its `session-wrapper` script which is expected to exec the passed command: https://wiki.ubuntu.com/LightDM#Adding_System_Hooks

NixOS takes control by setting the `session-wrapper` to [`services.xserver.displayManagar.session.script`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/lightdm.nix#L52), ie. the huge [`xsession`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/default.nix#L31) script.

The Exec command is by convention [`${cfg.displayManager.session.script} "${n}"`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/default.nix#L196), where `${n}` is the name of the desktop environment (optionally concatted with `+wm-name`).

So lightdm will normally launch something like this: `/nix/store/*xsession /nix/store/*xsession plasma5`. Which is reduced to the `/nix/store/*xsession plasma5` by [this line of code](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/default.nix#L57): 
```bash
      if [ "''${1:0:1}" = "/" ]; then eval exec "$1" "$2" ; fi
```

The `xsession` script will then extract the desktop environment name from the first argument and run the start script ([relevant code](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/default.nix#L160)).

#### Using upstream sessions

When using the upstream gnome.desktop session the above will simply reduce to executing `/nix/store/*gnome-session*/bin/gnome-session`, skipping `. /etc/profile` etc.

By applying this rather naive fix the gnome shell session is launched with at least $PATH etc. correctly set up (works with sddm too, which I think is an improvement #24298):
```diff
modified   nixos/modules/services/x11/display-managers/default.nix
@@ -32,6 +32,8 @@ let
     ''
       #! ${pkgs.bash}/bin/bash
 
+      . /etc/profile
+
       # Expected parameters:
       #   $1 = <desktop-manager>+<window-manager>
 
@@ -59,7 +61,6 @@ let
       # Now it should be safe to assume that the script was called with the
       # expected parameters.
 
-      . /etc/profile
       cd "$HOME"
 
       # The first argument of this script is the session type.
```

#### Actual solution?

The underlying problem seems to be that we're not using the session's `Exec` command to launch sessions as expected, making upstream sessions difficult. We're instead using [`session.start`](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/desktop-managers/plasma5.nix#L43) to do the actual launching through a big case statement in `xsession`.

We could create a proper `session-wrapper` which _only_ sets up the environment and then launches the session's `Exec` command. For legacy purposes the `Exec` command would be a reduced `xsession` script mostly containing a big case statement to select the correct session, ie. starting somewhere [around here](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/default.nix#L149).

This should allow us to handle both legacy and upstream sessions I think.

**Comment by worldofpeace** [comment]
@hedning Excellent!
I just need to meditate on it a few times to ensure proper absorption.

> This should allow us to handle both legacy and upstream sessions I think.

That is very much the outcome we want at least for now.

:+1:

**Comment by hedning** [comment]
Okay, I think I have something that could work okay. Tested in a VM that both Plasma5 and Gnome Shell launched correctly in LightDM, SDDM and GDM:

af8b14420d99f8c0d9a1fe1b2b07552d66224770: 
```
nixos/displayManager: Make it possible to use upstream sessions

We separate `xsession` into two parts, `xsessionWrapper` and `xsession`.
`xsessionWrapper` sets up the correct environment and then lauches the session's
Exec command (from the .desktop file).

`xsession` the _only_ handles launching desktop-managers/window-managers defined
through `services.xserver.desktopManager.session`.
```
and 
f6ab354cf0a220a662c4300c87597159e17eebe3:
```
nixos/gdm: Use common xsessionWrapper
```

**Comment by worldofpeace** [comment]
I've tested my not so secret project with this too ([elementary](https://elementary.io/)) and it works there as well.

It appears you've done (at least well enough) what you've set out to do and retain legacy support :)

**Comment by hedning** [comment]
> It appears you've done (at least well enough) what you've set out to do and retain legacy support :)

Yeah, mostly moved some code around so things should hopefully work the same as before.

On the GPaste/sessionPath issue; We could keep `gnome3.sessionPath` by adding the code to `displayManager.sessionCommands`. Checking against eg. `$XDG_CURRENT_DESKTOP` would keep the current behavior of only applying the `sessionPath` when launching a gnome session too. 

That should keep any potential breakage down, making it easier to land this PR? Does that make sense @jtojnar ?

**Comment by worldofpeace** [comment]
>On the GPaste/sessionPath issue; We could keep gnome3.sessionPath by adding the code to displayManager.sessionCommands. Checking against eg. $XDG_CURRENT_DESKTOP would keep the current behavior of only applying the sessionPath when launching a gnome session too.

That seems ok for now yet, we should just probably fix it instead of using `sessionPath`.

I personally still need `sessionPath` as I've ran into issues like with nautilus and gpaste a few times already, and don't actually have the time to fix them canonically.

**Comment by hedning** [comment]
Went through all the relevant tests and found some failures:

### xfce

Found and fixed a problem when using slim and `autoLogin` (which made the nixos xfce tests fail). Apparently the `sessionWrapper` (configured through [`login_cmd` in slim](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/x11/display-managers/slim.nix#L17) isn't passed any arguments when using autoLogin.

That's simply fixed by launching `xsession` from `xsessionWrapper` if it gets no arguments:

https://github.com/hedning/nixpkgs/commit/a0c090d447fe155b5ef9f9d8757b5f77a4e46a3b

(I squashed https://github.com/hedning/nixpkgs/commit/f6ab354cf0a220a662c4300c87597159e17eebe3)

### gnome3

The gnome3 test launches the `xterm` session since `desktopManager.default` gets set incorrectly. 

We need to register sessions added through  `extraSessionFilePackages` as valid session names when setting `desktopManager.default`. This probably means moving `extraSessionFilePackages` to `xserver.desktopManager` and checking the names there.

LightDM also checks the default session in an unnecessary way, so that needs a change. 

A working prototype (these tests pass: sddm.default sddm.autoLogin slim gnome3-gdm gnome3 plasma5 xfce xmonad i3wm):
- Allow default from extraSessionFilePackages: https://github.com/hedning/nixpkgs/commit/999ed1d6f2fec54d1cf77bdd510c73abac6b564a
- fix lightdm etc.: https://github.com/hedning/nixpkgs/commit/1995f44106a9b570fbee82226b6c94217fe1f33c

**Comment by hedning** [comment]
@jtojnar I can open a separate PR for the `xsession` split if that's preferable to cherry picking commits?

File: nixos/modules/services/x11/desktop-managers/default.nix [hunk 2]
```diff
@@ -91,9 +99,15 @@ in
         example = "none";
         description = "Default desktop manager loaded if none have been chosen.";
         apply = defaultDM:
+          let
+            extraNames = concatMap (extraPackage: concatMap (d: builtins.match "(.*).desktop" d)
+              (builtins.attrNames (builtins.readDir "${extraPackage}/share/xsessions")))
```
Comment (jtojnar) [inline nixos/modules/services/x11/desktop-managers/default.nix]: Build during evaluation is frowned upon for performance reasons, we will probably need to switch to a checking in later phase.

**Comment by hedning** [comment]
Pushed the `throw` to `trace` change, and added back `sessionPath` and `debug` implementations (did some rebasing and rewording): https://github.com/hedning/nixpkgs/commits/upstream-sessions

Feel free to sqaush or fixup any code as you see fit btw.

File: nixos/modules/services/x11/desktop-managers/gnome3.nix [hunk 3]
```diff
@@ -58,8 +58,12 @@ in {
       sessionPath = mkOption {
         default = [];
         example = literalExample "[ pkgs.gnome3.gpaste ]";
-        description = "Additional list of packages to be added to the session search path.
-                       Useful for gnome shell extensions or gsettings-conditionated autostart.";
+        description = ''
+          Additional list of packages to be added to the session search path.
+          Useful for GNOME Shell extensions or GSettings-conditional autostart.
+
+          Note that this should be a last resort, patching the package is preferred (see GPaste).
```
Comment (worldofpeace) [inline nixos/modules/services/x11/desktop-managers/gnome3.nix]: Indeed :smirk:

File: nixos/modules/services/x11/desktop-managers/gnome3.nix [hunk 4]
```diff
@@ -57,8 +57,12 @@ in {
...
```
Comment (benley) [inline nixos/modules/services/x11/desktop-managers/gnome3.nix]: Really minor punctuation nit:

"Note that this should be a last resort; patching the package is preferred (see GPaste)."
"Note that this should be a last resort, and patching the package is preferred (see GPaste)."

(either of the above is fine)

**Comment by jtojnar** [comment]
Thanks @hedning, now it looks like it is ready to go; all tasks completed and tests passing.

There is a space for more clean-ups, for example, removing pulseaudio from the session wrapper (it is supposed to be socket-activated on modern systems), but we can leave it for future PRs.

Asking maintainers of other DEs for their input: @ttuegel (KDE), @romildo (lxqt), @yegortimoshenko (xfce).

**Comment by hedning** [comment]
@worldofpeace @jtojnar We could check that the default session exists when actually building the `desktops` derivation perhaps?
https://github.com/NixOS/nixpkgs/blob/d9d92218ec53129a18740f385a3873973c95ec0d/nixos/modules/services/x11/display-managers/default.nix#L175?

**Pull request merged**